### PR TITLE
Update alloc after reconnect and enforece client heartbeat order

### LINKED ISSUE #12680: disconnected clients: Reconnecting allocation can cause a replacement that is immediately stopped
### Nomad version

Nomad 1.3.0-beta1

### Operating system and Environment details

Nomad Vagrantfile 
1 Server
2 Clients

### Issue

Creating a job that is constrained to a single node, can result in a second allocation being scheduled for that client being started and then immediately stopped when a disconnected client reconnects

### Reproduction steps

- `nomad run` the jobfile below that constrains by `node.unique.name`
- In a new session, start a watch on the job `watch nomad status <jobname>`
- Wait for the deployment to finish
- Start a watch for the job - you should see something similar to the following
- Simulate a network outage on the targeted client (e.g. `sudo iptables -I INPUT -s 192.168.56.11 -j DROP`)
- Wait for the allocation to transition to the unknown status
- Simulate a network reconnect for the targeted client (e.g. `sudo iptables -D INPUT -s 192.168.56.11 -j DROP`)
- When the allocation transition back to `running`, notice that a second allocation was started and immediately stopped
- Notice the original allocation did not stop

```shell
Allocations
ID                Node ID      Task Group  Version  Desired  Status       Created         Modified
2f8d72f2    94027db7   cache           0            stop       complete  2s ago            1s ago
043e25bd  94027db7   cache           0            run         running     1m44s ago    2s ago
```

#### Expected Result

The original allocation reconnects and no additional allocations are scheduled

#### Actual Result

An extra allocation was scheduled and then immediately stopped

### Job file (if appropriate)

```hcl
job "no-replace" {
  datacenters = ["dc1"]

  group "cache" {
    count = 1

    max_client_disconnect = "2m"

    constraint {
      attribute = "${node.unique.name}"
      value     = "nomad-client02"
      operator  = "="
    }

    network {
      port "db" {
        to = 6379
      }
    }

    task "redis" {
      driver = "docker"

      config {
        image = "redis:3.2"

        ports = ["db"]
      }

      resources {
        cpu    = 500
        memory = 256
      }
    }
  }
}
```

### LINKED ISSUE #14925: Job updates not applied to reconnected allocations
### Nomad version
Confirmed on 1.3.x, 1.4.x

### Operating system and Environment details

Ubuntu 20.04
3 servers
3 clients

### Issue

nomad job with `max_client_disconnect` becomes unmanageable  (no updates can be occur after successful updating job definition ) unless restarted, after job was in state "unknown" (nomad client was disconnected) 
 
### Reproduction steps

- Run the jobspec below
- Disconnect the nomad client where job is running
- After Job becomes in status `unknown` reconnect the nomad client
- After job s reconnected change the `image` to `httpd:latest` and run job again
- It will be successful but if you make curl request it still return nginx default response (allocation will be still using nginx container)
- If you look in the UI you can see that the Alloc that was placed on the disconnected/reconnected client does not get replaced and is running with the previous job version

#### Expected Result

Reconnected alloc will be replaced with newer jobspec config

#### Actual Result

Allocs that never transitioned to`unknown` get replaced, but the disconnected/reconnected alloc may not get replaced. In at least one manual bench test, I have seen it work as desired, but not deterministically.

### Job file (if appropriate)

```hcl
job "nginx-job" {
  datacenters = ["dc1"]

  group "nginx-group" {
    max_client_disconnect = "24h"

    count = 3

    spread {
      attribute = "${attr.unique.hostname}"
    }

    network {
      port "http" {
        to = 80
      }
    }

    task "nginx-task" {
      driver = "docker"

      config {
        // toggle these lines to create new job version
        image = "nginx:latest"
        // image = "httpd:latest"
        ports = ["http"]
      }

    }
  }
}
```

### Nomad Server logs (if appropriate)

Should have place 3 and stop 3

```shell
Oct 18 10:52:20 node-server-1 nomad[1819]:   results=
Oct 18 10:52:20 node-server-1 nomad[1819]:   | Total changes: (place 2) (destructive 0) (inplace 0) (stop 2) (disconnect 0) (reconnect 0)
```

### PR DESCRIPTION
**Pull request by lgfa29**
When a Nomad client with at least one alloc with [max_client_disconnect](https://developer.hashicorp.com/nomad/docs/job-specification/group#max-client-disconnect) misses a heartbeat the leader [updates its status to `disconnected`](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/heartbeat.go#L167-L173) which results in [evals for all jobs that have an allocation in that client](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/node_endpoint.go#L499-L500). This eval is reconciled by the scheduler, and [updates the alloc `ClientStatus` to `unknown` and appends a new state value](https://github.com/hashicorp/nomad/blob/v1.4.2/scheduler/reconcile.go#L1337-L1344) to indicate the allocation is considered disconnected.

When the client reconnects again [its status is set back to `ready`](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/node_endpoint.go#L489) since the heartbeat succeeds. The `ClientStatus` of the allocation is still `unknown` as it may have failed while the client was disconnected. This status will only be updated once the client calls the `Node.UpdateAlloc` against a server, which will [overwrite the `unknown` `ClientStatus` value set in the server](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/node_endpoint.go#L1247) with the correct status.

From the leader perspective this is what happens:

node-disconnect drawio

The problem described in #14925 happens because any eval after this disconnect/reconnect flow happens (such as a job update) is essentially indistinguishable from the last eval in the diagram above: the node is `ready` and the alloc `ClientStatus` is `running`, but these two scenarios need to be handle differently and so we need to store something in state to be able to detect this difference.

Current implementation uses the presence of a [`TaskClientReconnected`](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/structs/structs.go#L8156-L8157) task event to detect if [an alloc needs to reconnect](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/structs/structs.go#L10401-L10422), but this event is still present even after the alloc reconnects, so the scheduler always consider the alloc as still reconnecting.

While testing the fix for #14925 I would often hit #12680, which prevented the issue from being triggered since the extra alloc would not be considered as "reconnecting", so I included a fix for #12680 in this PR as well since their root cause is similar.

Clients use three main RPC methods to communicate its state with servers:

  - `Node.GetAllocs` reads allocation data _from the server_ and writes it _to the client_.
  - `Node.UpdateAlloc` reads allocation _from from the client_ and writes them _to the server_.
  - `Node.UpdateStatus` writes the client status to the server and is used as the heartbeat mechanism.

These RPC methods are called periodically by the client, and independently from each other. The usual mental model is that clients heartbeat first with `Node.UpdateStatus` and then update their allocation data with `Node.UpdateAlloc`, but this is not always true. If these two operations are reversed the diagram above will look like this:

node-disconnect drawio2

The state at the second last eval (`Node updates allocs`) looks just like the state when the node missed its heartbeat (`Node misses heartbeat`): `Node.Status: disconnected` and `Alloc.ClientStatus: running`.

This PR addresses these problems in the following ways:
- Don't rely on task events since only a limited number of them are kept in state 
- Record when a allocation reconnects as an `AllocState` entry to keep it consistent with how disconnects are recorded
- Check if the alloc needs to reconnect by verifying the last `AllocState`, it it's `unknown` the alloc reconnect has not been processed yet
- Enforce that nodes register and heartbeat successfully before being able to update their allocs. This makes the reconnect flow easier to reason about and more predictable.

The commits are split by different work chunks:

- 882904bfd04ae9cae81f1e3b3d9a91cde651059c is the main fix for #14925, both for system and service jobs.
- 720513f00214a3004b120ce41457252f1d0c21a8 fixes #12680 by properly handling `running` allocs in disconnected clients
- 956a3fa93ee9b34a02cb686c2add6ff48db37869 enforces clients to heartbeat before they can update their allocs. While 720513f00214a3004b120ce41457252f1d0c21a8 would be enough to fix #12680 I noticed other edge cases that I was not able to fix in the scheduler.

This is an alternative fix for #14948.

Closes #14925
Closes #12680

### CODE REVIEW

File: nomad/node_endpoint.go
Comment (lgfa29) [inline nomad/node_endpoint.go:1164]: I wasn't sure how to handle this, but I think it's fine to expect that all allocations being sent here are from the same client?

An alternative would be to just skip this check of there is no node ID.

File: nomad/structs/structs.go
Comment (lgfa29) [inline nomad/structs/structs.go:10410]: This assumes that states are order chronologically, which I think it's fine since entries are added using [`alloc.AppendState`](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/structs/structs.go#L10183-L10191)? Another option would be to traverse all values.

File: nomad/structs/structs.go
Comment (lgfa29) [inline nomad/structs/structs.go:10398]: I changed the return values here because the double boolean was a bit confusing and also kind of unnecessary since there's already a [`alloc.Expired`](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/structs/structs.go#L10354-L10382) method.

File: scheduler/reconcile_test.go
Comment (lgfa29) [inline scheduler/reconcile_test.go:5318]: It ended up a bit repetitive, but I think it's important to highlight that `AllocStates` is also used for scheduling decisions.

File: scheduler/reconcile_util_test.go
Comment (lgfa29) [inline scheduler/reconcile_util_test.go:899]: I switched to `assert` because these fields are all independent and it was annoying to find errors one by one 😅 . Now the test will print all failures at once.

File: nomad/node_endpoint.go
Comment (shoenig) [inline nomad/node_endpoint.go:1159]: 1.4?

File: nomad/structs/structs.go
Comment (shoenig) [inline nomad/structs/structs.go:10405]: does the reverse order matter? if so can you comment why

File: scheduler/reconcile_test.go
Comment (shoenig) [reply scheduler/reconcile_test.go:5318]: you could make a little helper function to tidy up the repetition

**Review by shoenig — APPROVED** [review]
Awesome investigation, @lgfa29! I just had small comments

File: nomad/node_endpoint.go
Comment (lgfa29) [reply nomad/node_endpoint.go:1159]: Ooops 😅

File: scheduler/reconcile_test.go
Comment (lgfa29) [reply scheduler/reconcile_test.go:5318]: Ah yeah, it was actually simpler to just create a variable with the most common value. Thanks!

**Comment by lgfa29** [comment]
Ops, I found a regression while doing some extra testing where the alloc keeps running when the job is stopped while the client is still disconnected. I'm still investigating it.

**Comment by lgfa29** [comment]
> Ops, I found a regression while doing some extra testing where the alloc keeps running when the job is stopped while the client is still disconnected. I'm still investigating it.

More race condition fun: f5ce8a96b3b9a7676c23fba78faeb583988a364a 😄 

This time between alloc reads and writes in the client. If the client reads the alloc state before writing it, it will trigger the `Reconnect` flow even if `DesiredStatus: stop`. While this was existing bug, by delaying the alloc status write this PR just made this problem more likely to happen.

File: nomad/structs/structs.go
Comment (schmichael) [inline nomad/structs/structs.go:954]: Hm I'd really like to avoid more conditionally-initialized fields as this is exactly what makes figuring out things like `stripped allocs sent from clients` so hard.

AFAICT this struct is used for *three mostly disjoint requests:*

1. Stripped allocs from clients Node.UpdateAlloc RPC using the Allocs, NodeID (new), and WriteRequest fields
2. Raft log message using the Allocs, Evals, and WriteRequest fields
3. Plan updates using the AllocsStopped, AllocsUpdated, and Job fields

1 & 2 using the same alloc is defensible but 3 just makes no sense these days. 3 isn't your problem though at all.

I can't think of a way to change this that I'd be comfortable backporting, so maybe let's put a big ugly comment here and file an issue to clean this up someday? I really don't want to hold up your PR with refactoring this mess.

File: scheduler/reconcile.go
Comment (schmichael) [inline scheduler/reconcile.go:1206]: Is this alloc a new alloc to be applied via a plan? Just making sure we're not mutating an object that's in the statestore.

**Review by schmichael — COMMENTED** [review]
In the description:

> When a Nomad client misses a heartbeat the leader [updates its status to disconnected](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/heartbeat.go#L167-L173)

should probably read:

> When a Nomad client _with at least one alloc with [`max_client_disconnect`](https://developer.hashicorp.com/nomad/docs/job-specification/group#max-client-disconnect)_ misses a heartbeat...

Clients without an alloc with `max_client_disconnect` set will transition to `down` and allocs to `lost`. Minor pedantic point, but I suspect we may be linking to this PR in the future to explain this behavior so I think it's worth being precise. :grin: 

> Don't rely on task events since only a limited number of them are kept in state

:tada: I'm really happy to be keeping task events for humans only.

File: scheduler/reconcile_util.go
Comment (schmichael) [inline scheduler/reconcile_util.go:243]: filterByTainted has `time.Now()` passed in to help with testing. 
```suggestion
				expired = alloc.Expired(now)
```

File: scheduler/reconcile_util.go
Comment (schmichael) [inline scheduler/reconcile_util.go:268]: I thought this was prevented by the node readiness check added in Node.UpdateAlloc?

File: scheduler/util.go
Comment (schmichael) [inline scheduler/util.go:200]: Same question as above: I thought this was prevented in the node endpoint changes?

Since the eval's WaitIndex is _at least_ its CreatedIndex, and it will have been created by the node heartbeat at the earliest, I don't think there's any chance of a "the statestore might not have applied the raft log the scheduler needs to make the right decision yet" issue, so as long as the node endpoint prevents this logic I'm not sure why we also need to prevent it here?

**Review by schmichael — APPROVED** [review]
Absolutely fantastic work from the repro to the description to the code! What a complex situation to debug and fixup.

An e2e test would be nice since there's so much emergent nondeterministic behavior at play. Your unit tests are great but come with some assumptions about the state of the data being tested that may not cover everything that can happen at runtime.

I forget if our `Test{Agent,Server}` make it possible to write an integration test which at least exercises these changes with a real `client.Client`.

The obvious problem with an e2e testing is how to simulate a disconnected client. Do I have a terrible idea for you! :sweat_smile: 

Schedule count=2 nomad agents via raw exec with the "fake" nomad's datacenter = `your-test-name`, and then schedule a job with `max_client_disconnect` set that targets that datacenter.

Grab the node ID of which of the 2 your alloc got scheduled on, send it a kill signal, let it restart via the parent Nomad, and then assert this intended behavior.

I think the nested-nomad agent should work, and the unique datacenter trick should prevent any flakiness from other jobs colliding with your careful placement scheme.

**Does _not_ block merging this though.** Probably best as a followup.

File: scheduler/reconcile_util.go
Comment (lgfa29) [reply scheduler/reconcile_util.go:268]: Yup, the RPC update is enough for this particular issue, but I was wondering if there was another scenario that we don't know about that could trigger this condition (like some data corruption? or bad alloc in state?) and so I kept the reconciler fix as well as a kind of layered defense. 

But I can remove this as well, less code to think about? 😅

File: scheduler/reconcile.go
Comment (lgfa29) [reply scheduler/reconcile.go:1206]: Nice catch! I don't think it's being copied. Tracing it back, this `alloc` comes from these [`reconnecting[alloc.ID] = alloc`](https://github.com/hashicorp/nomad/blob/v1.4.2/scheduler/reconcile_util.go#L249). 

It's so easy to break stuff 😅

File: .changelog/15068.txt
Comment (lgfa29) [reply .changelog/15068.txt:6]: Yeah, makes sense. I didn't know how to call these either. I probably called them disconnected allocs in code comments as well 🤷

File: nomad/node_endpoint.go
Comment (lgfa29) [reply nomad/node_endpoint.go:1163]: I tried to use `n.ctx.NodeID` but you're right, it's not present when the request is forwarded. I think it would take quite a bit of refactoring since we would need to pass it in the `n.src.forward` call and that affects all endpoints.

Another option would be to add it to [`InternalRPCInfo`](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/structs/structs.go#L228-L233) before forwarding the request, but I'm not sure if that's a good 😅

**Review by tgross — COMMENTED** [review]
I haven't finished reviewing yet but wanted to surface my comments before I wrap up for the day 😀 . Will pick this back up first thing tomorrow.

File: scheduler/reconcile_util.go
Comment (lgfa29) [reply scheduler/reconcile_util.go:268]: That's what I meant by the layered approach. The RPC change fixes things externally but without this change the reconciler can still _in theory_  generate an invalid result given a these specific conditions. 

I would've liked to be able to fix everything in the scheduler, but I found one case for system jobs that I couldn't and only the RPC ordering fixed it, and I suspect there may be other edge cases like this, so I kept both approaches.

But I agree that a comment stating this should not happen would be good.

**Comment by lgfa29** [comment]
I think I found a bug in the new `NeedsToReconnect` method that is preventing an alloc to disconnect twice. I will investigate further tomorrow.

File: nomad/node_endpoint.go
Comment (schmichael) [reply nomad/node_endpoint.go:1165]: I think if you just grab it before calling `forward` it would work, but that's a bit odd as forward is almost always the first thing we do.

What if we just dropped the field and always plucked it off the alloc? That would avoid adding another only-used-in-1-of-3-places field to the overloaded request struct. We already always have it on the stripped allocs, so we'd finally make that decision sensible.

We can ship it with the new NodeID field as its not worth holding up these excellent fixes for.

Comment (lgfa29) [reply nomad/node_endpoint.go:1165]: > I think if you just grab it before calling `forward` it would work

The problem with this is that the leader would then need to skip the validation, or more specifically, a server would need to ignore the check if the request was forward and assume that the check has been done at some point. 

I thought this would be a bigger problem for cross-region request since the first server would not find the node, but now that I think about it, it's kind of weird to have a cross-region request in this endpoint 😅 

> What if we just dropped the field and always plucked it off the alloc? 

Yeah, that sounds good to me. I wasn't sure which way to go any way and this is probably the most reliable option without adding more code 👍

File: scheduler/reconcile_test.go
Comment (tgross) [inline scheduler/reconcile_test.go:5278]:
```suggestion
	// Create 2 allocs and simulate one have being previously disconnected and then reconnected.
```

I think, based on these alloc states?

**Review by tgross — APPROVED** [review]
LGTM! (Once the docstrings are fixed up as being discussed)

File: nomad/node_endpoint.go
Comment (lgfa29) [reply nomad/node_endpoint.go]: Accessing Node ID from allocs will be the default:
https://github.com/hashicorp/nomad/pull/15068#discussion_r1013503431

File: nomad/structs/structs.go
Comment (lgfa29) [reply nomad/structs/structs.go]: Removed the new field as per suggestion [above](https://github.com/hashicorp/nomad/pull/15068#discussion_r1013503431). Thanks!

File: scheduler/reconcile_util.go
Comment (lgfa29) [reply scheduler/reconcile_util.go]: I'm sold now 😄 

Adding code to the scheduler that we don't expect to hit just add stuff to an already complex part of the code, so I removed this check and its equivalent for `system` scheduling. Thanks for the feedback!

File: scheduler/reconcile_test.go
Comment (lgfa29) [reply scheduler/reconcile_test.go:5278]: Oh that's right, good catch!

**Pull request merged**
